### PR TITLE
Hw0b

### PR DESCRIPTION
**Pull request by tobytoyuito**

### CODE REVIEW

File: HW0/P4/P4.py [hunk 1]
```diff
@@ -30,47 +30,92 @@
...
     # Read the observation array and plot it (Part 2)
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    m = np.loadtxt("P4_measurements.txt", delimiter=",")
+    C = np.array([[1./rx, 0.0, 0.0],
+                  [0.0, 1./ry, 0.0],
+                  [0.0, 0.0, 1./rz]])
+    s_obs=np.dot(C, m.T)
```
Comment (samighoche) [inline HW0/P4/P4.py]: Nice! matrix operations are faster than vector operations in a for loop

I would be careful with naming this matrix C though because C in the context of this problem is actually another matrix

File: HW0/P4/P4.py [hunk 2]
```diff
@@ -30,47 +30,92 @@
...
+    sk[:,0]=s0
+    s = s0
+    Sigma = Sigma0
+    for i in range(1,K):
+        s_tilt = np.dot(A, s) + a
+        Sigma_tilt = np.linalg.inv(np.dot(A, np.dot(Sigma, A.T)) + np.dot(B, B.T))
+        Sigma = np.linalg.inv(Sigma_tilt + np.dot(C.T, C))
+        s = np.dot(Sigma, (np.dot(Sigma_tilt, s_tilt)+ np.dot(C.T, m.T[:,i])))
+        sk[:, i] = s
+
```
Comment (samighoche) [inline HW0/P4/P4.py]: nice! very compact and succinct:)

File: HW0/P5/P5.txt [hunk 3]
```diff
@@ -1 +1,10 @@
...
+   2 bags: It is similar to 1 bag but save 1 seconds in distributing bags and 1 seconds to sum up bags from 1 bag in two cashiers. There fore, it takes 25-1-1 = 23 seconds. 
+   There is a better strategy. First, I should give all bags to a cashier and this takes 1 second. And then this cashier can give 128 bags to another cashier and this takes another second. Every time the cashier do this, it takes 1 second. Now there is a problem: when to stop giving bags and count bags themselves. The answer is when they have 4 bags, they should count them themselves. If cashier give it to other, it takes 1 second to give it to another cashier, 1 second to sum two bags, 1 second to communicate and 1 second to sum two number, so it would be 4 seconds. However, if cashier counts himself, it just takes 3 seconds. 
+To sum up, they should give half of the bags in their hand to another cashier until they have 4 bags in hand and this strategy takes 7+3+6*2=22 seconds.
```
Comment (samighoche) [inline HW0/P5/P5.txt]: I agree with all your answers except 6 where I found it hard to follow your argument, especially for 1bag, 2 bags

File: HW0/P6/P6B.py [hunk 4]
```diff
@@ -16,16 +16,26 @@ def burnTime(t):
 
     # Use a variety of wait times
     ratio = []
-    wait_time = []
+    wait_time = [10.0 ** ii for ii in range(-6, 0)]
```
Comment (samighoche) [inline HW0/P6/P6B.py]: you might want to have more data points. A way to do this without too much manual work or double for loops is to use 
np.logspace(-6, 0) (gives 50 points by default), or np.logspace(-6, 0, x) where x is the total number of points you want

File: HW0/P5/P5.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import math
+
+def fx(N):
+    return math.ceil(np.log(N))
```
Comment (georgelok) [inline HW0/P5/P5.py]: Good and important use of ceil. :)

File: HW0/P5/P5.txt [hunk 6]
```diff
@@ -1 +1,10 @@
-# Your answers here.
+1. 255 seconds.
+2. Yes. There are 8 people totally. They can distribute 32 bags to one person and each person only count their own bags. 31+3=34 seconds. In this case, they just need 34 seconds to do it.
+3. If there are infinite number of people. They can use divide and conquer algorithm. They can use 128 people to sum each 2 bags first. And then sum them two by two at the same time and continue again until we have one sum. This strategy takes log(256)=8 seconds.
+4. If we have N bags. The fastest time to count them would be ceil(log2(N))
```
Comment (georgelok) [inline HW0/P5/P5.txt]: I agree with this answer, but it would have been nice to have a little more discussion on why ceil is needed.

File: HW0/P6/P6B.py (hunk 4 above)
Comment (georgelok) [inline HW0/P6/P6B.py]: importing numpy and using np.logspace(-6, 0) can simplify this.

File: HW0/P4/P4.py [hunk 7]
```diff
@@ -30,47 +30,92 @@
...
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    A = np.array([[1.0, 0.0, 0.0, dt, 0.0, 0.0],
```
Comment (georgelok) [inline HW0/P4/P4.py]: np.diag() (and other numpy functions) can make making these matricies a little more clean.

Of course, it's very clear what this matrix is in this form, so just giving a heads up. :)

File: HW0/P5/P5.txt (hunk 3 above)
Comment (adisuissa) [inline HW0/P5/P5.txt]: Indeed it seems like you misunderstood part 6 of the question.
In each second, either 1, 2, or X bags can be passed, and you cannot pass all bags in 1 second (or 8).
